---
title: "Martha_R_Assignment"
author: "Martha"
date: "10/13/2019"
output: html_document
---

PART I: Replicating the UNIX assignment in R

Step 1: Downloading data files from github through the class website
```{r}

download.file("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2019/master/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", destfile = "./fang_et_al_genotypes.txt")

download.file("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2019/master/assignments/UNIX_Assignment/snp_position.txt", destfile = "./snp_position.txt")
```

Step 2: Loading the data in R
```{r}
library(tidyverse) #Comment:loading the required package

#Reading in the fang et al dataset. #here, we are giving the data file a name using the sign <- name on the left and dataset on the right
fang <- read_tsv("./fang_et_al_genotypes.txt") 

#Reading in the snp dataset. #here, we are giving the data file a name using the sign <- name on the left and dataset on the right
snps <- read_tsv("./snp_position.txt")
```

Step 3: Data inspection
```{r}
dim(fang) #this line allows us to see how may rows and columns are in the fang_et_al_genotypes dataset
dim(snps) #this line allows us to see how may rows and columns are in the snp_positions dataset
fang[1:10, 1:5] #Because the columns are too many, this code allows us to see only 10 rows and 5 columns of the fang dataset
snps[1:10, 1:5] #Because the columns are too many, this code allows us to see only 10 rows and 5 columns of the snp dataset

str(fang) #Prints details of the fang dataset
str(snps) #Prints details of the snp dataset

summary(fang) ##Prints details of the fang dataset
summary(snps) #Prints details of the snp dataset

unique(fang$Group)# prints all the groups in the fang dataset
```

Step 4: Data processing (obtaining the needed groups and transposition)
```{r}
maize_transposed <- filter(fang, Group %in% c("ZMMIL", "ZMMLR", "ZMMMR")) %>% 
  select(-JG_OTU, -Group) %>% column_to_rownames(., var = "Sample_ID") %>% t() %>% 
  as.data.frame() %>% rownames_to_column(., var = "SNP_ID")
#The above code is a loop that is used to create a processed file called "maize_transposed" by following the respective steps.

teosinte_transposed <- filter(fang, Group %in% c("ZMPBA", "ZMPIL", "ZMPJA")) %>% 
  select(-JG_OTU, -Group) %>% column_to_rownames(., var = "Sample_ID") %>% t() %>% 
  as.data.frame() %>% rownames_to_column(., var = "SNP_ID") 
#The above code is a loop that is used to create a processed file called "maize_transposed" by following the respective steps.

snps_formatted <- snps %>% select(SNP_ID, Chromosome, Position) #this formats the snp dataset as needed

maize_merged <- merge(snps_formatted, maize_transposed ) %>%
  filter(., Position != "multiple",  Chromosome != "unknown") #this merges the formatted snps file and the transposed maize dataset
teosinte_merged <- merge(snps_formatted, teosinte_transposed) %>%
  filter(., Position != "multiple",  Chromosome != "unknown")
 #this merges the formatted snps file and the transposed teosinte dataset
```


Step 5: Replacing ?/? with -/- 
```{r}
maize_replaced <- maize_merged %>% as.matrix() #Creates a file named "maize_replaced" which is set up as a matrix
teosinte_replaced <- teosinte_merged %>% as.matrix() #Creates a file named "teosinte_replaced" which is set up as a matrix

maize_replaced[maize_replaced == "?/?"] <- "-/-" #Replaces ?/? with -/- in the maize_replaced file 
teosinte_replaced[teosinte_replaced == "?/?"] <- "-/-" #Replaces ?/? with -/- in the teosinte_replaced file 
maize_replaced <- as.tibble(maize_replaced) #add comment
teosinte_replaced <- as.tibble(teosinte_replaced)
```

Step 6: Extracting all information for all 10 maize chromosomes
```{r}
for (i in 1:10) {
 
  m1 <- maize_merged %>% filter(., Chromosome == i[1]) %>% arrange(.,as.numeric(Position))
  m2 <- maize_replaced %>% filter(., Chromosome == i[1]) %>% arrange(.,desc(as.numeric(Position)))
 
  write_tsv(m1, paste("maize_chr", i[1], "increasing.txt", sep = "_"))
  write_tsv(m2, paste("maize_chr", i[1], "decreasing.txt", sep = "_"))
 
  t1 <- filter(teosinte_merged, Chromosome == i[1]) %>% arrange(., as.numeric(Position))
  t2 <- filter(teosinte_replaced, Chromosome == i[1]) %>% arrange(., desc(as.numeric(Position)))
 
  write_tsv(t1, paste("teosinte_chr", i[1], "_increasing.txt", sep = "_"))
  write_tsv(t2, paste("teosinte_chr", i[1], "_decreasing.txt", sep = "_"))
}
```



PART II: Using ggplot to visualize our data

Step 1: transposing the fang original dataset
```{r}
fang_transposed <- fang %>%  select (-JG_OTU, -Group) %>% column_to_rownames(., var = "Sample_ID") %>% 
  t() %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID") #this code transposes the entire fang et al dataset without selecting specific columns
```


Step 2: Merging the transposed file with the snps file
```{r}
fang_merged <- merge(snps_formatted, fang_transposed ) %>% 
  filter(., Position !="multiple", Position != "unknown")
```

Step 3: Generating graphs
```{r}
ggplot(fang_merged, aes(x=as.character(Chromosome))) + geom_bar(aes(fill = as.character(Chromosome))) #used to establish how many SNPs are in each chromosome.

ggplot(fang_merged, aes(x= as.numeric(Position))) + geom_density(aes(fill = as.character(Chromosome))) + facet_wrap(~ as.character(Chromosome), nrow = 2, ncol = 5) + theme_bw()
```

Step 4:loading tidyverse
```{r}
library (tidyverse)
```

Step 5: Tidying up the fang dataset
```{r}
tidy_fang <- fang %>% select(-JG_OTU) %>% 
  gather(key = "SNP_ID", "Sequence", names(fang) [4:986] )
```

Step 6: Missing data and amount of heterozygosity
```{r}
tidy_fang <- tidy_fang %>%
  mutate(new_sequence = ifelse(Sequence %in% c("A/A","T/T","C/C","G/G"), "Homozygous",
  ifelse(Sequence == "?/?", "Missing","Heterozygous")))
```

Step 7: Creating a bar graph
```{r}
ggplot(tidy_fang, aes(x = Sample_ID, fill = new_sequence)) +
  geom_bar(position = "fill") + theme_bw()
```

Step 8: Creating graph after normalizing the data
```{r}
ggplot(tidy_fang, aes(x = Group , fill = new_sequence)) + geom_bar(position = "fill") +
  theme_bw() + theme(axis.text.x = element_text(angle = 90))
```

Step 9: Creating my own graph
```{r}
ggplot(filter(tidy_fang, Sequence != "?/?") , aes(x = Sample_ID, fill = Sequence)) +
  geom_bar(position = "fill") + theme_bw()
```

